### PR TITLE
add streams extension

### PR DESCRIPTION
**Pull request by Acconut**

### CODE REVIEW

**Comment by qsorix** [comment]
I like it. A few comments below but if you disagree with them, I'm happy to go plus-one already.

One major comment. If there will be a header to advertise supported extensions (`TUS-Extensions` as #48 suggests), then each extension must have an official name to use. So I would add sth like

```
If this extension is supported on the server, then server's response to OPTIONS
request will include:
  TUS-Extensions: streams
````

Minor stuff follows.

Wouldn't it help server-side if the client explicitly declared it is indeed using `streams` extension? Otherwise it is hard to say if the client is buggy or the extension is being used. In a situation when the client is buggy, server resources would be wasted for an extended period of time. Client could finish an upload with server still waiting for that Entity-Length and more data. And I'm assuming this extension will be used for really big uploads, so wasted disk space may be significant.

> Once the total size of the entire upload
is known it MUST be included as the `Entity-Length` header's value in the next
`PATCH` request.

I would rephrase it to skip the word "next". Server will be happy if the `Entity-Length` is in at least one request. Client can send it once it knows it, or in the last PATCH request. I imagine a sensible implementation would send it in *every* request once it is known.

Should it be said explicitly that the server MUST reject the upload if it receives two different values for the Entity-Length? Or we don't care? Or server can not care and use the first value it received?

> Content-Length: 0

Quoted from the first example request. I don't think it is of any importance in this scope. I think the examples should either list all required headers (e.g. include `TUS-Resumable`) or list only the ones that matter in each case. So here, instead of a header, there should be a comment highlighting the fact that the `Entity-Length` is missing.

**Comment by Acconut** [comment]
> One major comment. If there will be a header to advertise supported extensions (TUS-Extensions as #48 suggests), then each extension must have an official name to use.

I planned to add this once #48 is merged and we're sure the TUS-Extension header doesn't change.

> Wouldn't it help server-side if the client explicitly declared it is indeed using streams extension?

Good point, what you think about using `Entity-Length: streaming` in the initial `POST` request?

> Should it be said explicitly that the server MUST reject the upload if it receives two different values for the Entity-Length? Or we don't care? Or server can not care and use the first value it received?

Another good point, for ease of use I would recommend to use the last value sent (no worrying about which status code to return). This also allows uploads when the resource's size unexpectedly grew.

> Quoted from the first example request. I don't think it is of any importance in this scope. I think the examples should either list all required headers (e.g. include TUS-Resumable) or list only the ones that matter in each case. So here, instead of a header, there should be a comment highlighting the fact that the Entity-Length is missing.

I will remove `Content-Length`. `TUS-Resumable` wasn't added since I wanted to wait until #48 is merged (see my first sentence).

**Comment by kvz** [comment]
> Another good point, for ease of use I would recommend to use the last value sent (no worrying about which status code to return). This also allows uploads when the resource's size unexpectedly grew.

How about using the biggest `Entity-Length` received, vs. the latest?

**Comment by qsorix** [comment]
> How about using the biggest Entity-Length received, vs. the latest?

All good ideas.

But... should it really be possible to send different values? If a client is capable of extending the streamed data, it knows about it. So it should not send `Entity-Length` at all until it is sure there won't be more data. This covers the cases when more data appear.

But let's say you're uploading live video footage that is supposed to last exactly 30 minutes and the codecs used allow you to predict it's size. You declare `Entity-Length` in advance but then suddenly somebody presses 'Stop' earlier. Now you want to say that you're actually done, but you need to send smaller `Entity-Length`.

In both cases, client-side developer actually knows a change to Entity-Length is possible and can refrain from sending it sooner than it knows the final value for sure.

To summarize, I think the value, once present, must never change. But to free server-side implementations from the burden of checking that, I would say that it must never change, and if it does, the behaviour is implementation-defined (or undefined).

> Good point, what you think about using Entity-Length: streaming in the initial POST request?

I'm fine with that. Although I would slightly prefer a generic mechanism to declare extensions in use. Eventually we may have an extension that cannot be implicitly enabled, and at that point it would make more sense to enable all extensions in the same manner.

**Comment by Acconut** [comment]
> To summarize, I think the value, once present, must never change.

Seems to be the best solution. But I think the behaviour when receiving a second `Entity-Length` should be defined. I vote for simply ignoring it.

> Although I would slightly prefer a generic mechanism to declare extensions in use. Eventually we may have an extension that cannot be implicitly enabled, and at that point it would make more sense to enable all extensions in the same manner.

Currently this doesn't make a lot of sense since extension who need to be announced to the server include their own header or values.

**Comment by Acconut** [comment]
This extension would be finished now.

**Comment by kvz** [comment]
Added a few nitpicking commits, otherwise LGTM
